java/ql/src/semmle/code/java/security/ExternalAPIs.qll

```diff
@@ -0,0 +1,146 @@
+/**
+ * Definitions for reasoning about untrusted data used in APIs defined outside the
+ * database.
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+
+module ExternalAPIs {
+  /**
+   * A `Method` which is considered a "safe" external API from a security perspective.
+   */
+  abstract class SafeExternalAPIMethod extends Method { }
+
+  /** The default set of "safe" external APIs. */
+  class DefaultSafeExternalAPIMethod extends SafeExternalAPIMethod {
+    DefaultSafeExternalAPIMethod() {
+      this instanceof EqualsMethod
+      or
+      getName().regexpMatch("size|length|compareTo|getClass|lastIndexOf")
+      or
+      this.getDeclaringType().hasQualifiedName("org.apache.commons.lang3", "Validate")
+      or
+      getQualifiedName() = "Objects.equals"
+      or
+      getDeclaringType().getQualifiedName() = "java.lang.String" and getName() = "equals"
+      or
+      getDeclaringType().hasQualifiedName("com.google.common.base", "Preconditions")
+      or
+      getDeclaringType().getPackage().getName().matches("org.junit%")
+      or
+      getDeclaringType().hasQualifiedName("com.google.common.base", "Strings") and
+      getName() = "isNullOrEmpty"
+      or
+      getDeclaringType().hasQualifiedName("org.apache.commons.lang3", "StringUtils") and
+      getName() = "isNotEmpty"
+      or
+      getDeclaringType().hasQualifiedName("java.lang", "Character") and
+      getName() = "isDigit"
+      or
+      getDeclaringType().hasQualifiedName("java.lang", "String") and
+      getName().regexpMatch("equalsIgnoreCase|regionMatches")
+      or
+      getDeclaringType().hasQualifiedName("java.lang", "Boolean") and
+      getName() = "parseBoolean"
+      or
+      getDeclaringType().hasQualifiedName("org.apache.commons.io", "IOUtils") and
+      getName() = "closeQuietly"
+      or
+      getDeclaringType().hasQualifiedName("org.springframework.util", "StringUtils") and
+      getName().regexpMatch("hasText|isEmpty")
+    }
+  }
+
+  /** A node representing data being passed to an external API. */
+  class ExternalAPIDataNode extends DataFlow::Node {
+    Call call;
+    int i;
+
+    ExternalAPIDataNode() {
+      (
+        // Argument to call to a method
+        this.asExpr() = call.getArgument(i)
+        or
+        // Qualifier to call to a method which returns non trivial value
+        this.asExpr() = call.getQualifier() and
+        i = -1 and
+        not call.getCallee().getReturnType() instanceof VoidType and
+        not call.getCallee().getReturnType() instanceof BooleanType
+      ) and
+      // Defined outside the source archive
+      not call.getCallee().fromSource() and
+      // Not a call to an method which is overridden in source
+      not exists(Method m |
+        m.getASourceOverriddenMethod() = call.getCallee().getSourceDeclaration() and
+        m.fromSource()
+      ) and
+      // Not already modelled as a taint step
```

Comment:
`.qlhelp` above used "modeled" (single "l"). It appears this project is not consistent with this, but it would probably good to at least be consistent for files related to each other, as it is the case here.